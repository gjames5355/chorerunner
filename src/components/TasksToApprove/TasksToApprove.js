import React from 'react';
import ApiService from '../../services/api-service';
import HouseholdContext from '../../contexts/HouseHoldContext';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faThumbsUp, faThumbsDown } from '@fortawesome/free-solid-svg-icons';

export default class TasksToApprove extends React.Component {
  state = {
    tasks: [],
  };
  static contextType = HouseholdContext;
  componentDidMount() {
    ApiService.getTasksToApprove(this.props.householdId).then(result => {
      this.setState({
        tasks: result,
      });
    });
  }

  handleUpdateTaskStatus = (taskId, householdId, status, points, memberId) => {
    ApiService.parentUpdateTaskStatus(
      taskId,
      householdId,
      status,
      points,
      memberId
    ).then(result => {
      let updatedTasksToApprove = this.state.tasks.filter(
        task => task.id !== taskId
      );
      this.setState({
        tasks: updatedTasksToApprove,
      });
      let tasks = this.context.tasks;
      if (status === 'approved') {
        let memberTaskList = tasks[memberId];
        let filteredTasks = memberTaskList.tasks.filter(task => {
          return task.id !== taskId;
        });
        tasks[memberId].tasks = filteredTasks;
        tasks[memberId].total_score += points;
        this.context.setTasks(tasks);
      }
      if (status === 'assigned') {
        tasks[memberId].tasks.map(task => {
          if (task.id === taskId) {
            task.status = 'assigned';
          }
          return null;
        });
        this.context.setTasks(tasks);
      }
    });
  };

  render() {
    const { householdId, memberId } = this.props;
    return (
      <section className='tasks-to-approve'>
        <h3>Completed Tasks</h3>
        <ul className='householdpage-member-task-list'>
          {this.state.tasks.map(task => {
            if (task.member_id === memberId) {
              return (
                <li key={task.id} className='completed-task'>
                  <div className='completed-task-container'>
                    <div className='hhp-title'>
                      <span>{task.title}</span>
                    </div>
                    <div className='hhp-points'>
                      <span>points: {task.points}</span>
                    </div>
                  </div>
                  <div className='thumb-container'>
                    <button
                      className='thumb'
                      onClick={() =>
                        this.handleUpdateTaskStatus(
                          task.id,
                          householdId,
                          'approved',
                          task.points,
                          task.member_id
                        )
                      }
                    >
                      <FontAwesomeIcon
                        icon={faThumbsUp}
                        size='lg'
                        color='rgb(255, 217, 0)'
                      />
                    </button>
                    <button
                      className='thumb'
                      onClick={() =>
                        this.handleUpdateTaskStatus(
                          task.id,
                          householdId,
                          'assigned',
                          task.points,
                          task.member_id
                        )
                      }
                    >
                      <FontAwesomeIcon
                        icon={faThumbsDown}
                        size='lg'
                        color='rgb(255, 217, 0)'
                      />
                    </button>
                  </div>
                </li>
              );
            } else {
              return null;
            }
          })}
        </ul>
      </section>
    );
  }
}
