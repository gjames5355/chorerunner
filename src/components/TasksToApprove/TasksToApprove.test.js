import React from 'react';
import ReactDOM from 'react-dom';
import TasksToApprove from './TasksToApprove';
import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';

const tasks = [
  { title: 'Obtain currency', id: 17, points: 3, status: 'assigned' },
];

it('TasksToApprove renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<TasksToApprove tasks={tasks} />, div);
  ReactDOM.unmountComponentAtNode(div);
});

it('renders Tasks To approve as expected', () => {
  const tree = renderer
    .create(
      <MemoryRouter>
        <TasksToApprove tasks={tasks} />
      </MemoryRouter>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
