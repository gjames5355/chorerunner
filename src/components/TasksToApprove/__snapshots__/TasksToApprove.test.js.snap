// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Tasks To approve as expected 1`] = `
<section
  className="tasks-to-approve"
>
  <h3>
    Completed Tasks
  </h3>
  <ul
    className="householdpage-member-task-list"
  />
</section>
`;
